### DISCUSSION TITLE
Help for a Python Novice? - TESLA Tokens

### DISCUSSION THREAD

**Discussion by Pyromancer579** [timdorr/tesla-api#295] [Q&A] [answered]
I'm not a coder, it was called programming when I last did it in properly back in the 1980's!

So please take pity on me.

I have a Raspberry Pi and learned a bit of Python last year so that I could use the code below to switch my Tesla Powerwall from whatever mode it was in (C_MODE - usually 'Self Powdered') to 'Backup' and visa versa.

My system stopped working a week ago when Tesla updated their authentication process.

Can anyone provide some Python code to replace the code I was using and get my system back up and running, ideally something that I can simply cut and paste into my existing code?

Here are the basics of what I was using, but this doesn't work anymore:

import requests
from requests.auth import HTTPDigestAuth
import datetime
import time
reply=''
count=0
import sys
import asyncio
if sys.version_info[0] == 3 and sys.version_info[1] >= 8 and sys.platform.startswith('win'):
    asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
from tesla_api import TeslaApiClient

async def Starter():
    global C_MODE
    global reserve
    global PWCharge
    client = TeslaApiClient(TES_ID, TES_PW)
    energy_sites = await client.list_energy_sites()
    assert(len(energy_sites)==1)
    reserve = await energy_sites[0].get_backup_reserve_percent()
    await client.close()

async def Boost():
    client = TeslaApiClient(TES_ID, TES_PW)
    energy_sites = await client.list_energy_sites()
    assert(len(energy_sites)==1)
    N_MODE = await energy_sites[0].get_operating_mode()
    if N_MODE==C_MODE:
        await energy_sites[0].set_operating_mode('backup')
        N_MODE = await energy_sites[0].get_operating_mode()
    await client.close()

I've read through the discussions and issues 215/260 and don't really understand what I have read, so I'm hoping that someone out there can help me get back to using my Powerwall efficiently and effectively again.

**Comment by ralfonso4** [comment, marked as answer]
Seems like you have to update your TeslaApiClient with the new OAuth2.0
Find the part with the autentication process.
I think included here is the answer: https://github.com/timdorr/tesla-api/discussions/258#discussioncomment-338593

Beeing a "Old rusty hippie programer" is no excusion! ;-)

**Reply by Pyromancer579** [reply]
Thanks.

I had already tried that, it runs without errors, but I don't understand what it is doing or how to use whatever it is doing to allow me to control my Powerwall.

Any guidance would be much appreciated.

**Reply by Pyromancer579** [reply]
Okay, have started to get my head around the code from #258 (liberal application of WD-40 to rusty programming brain) and have now slotted it into my existing code, it almost works, I get:

**GET authorization form successful after 1 attempt(s).** - sometimes 2 attempts.

Which is a good start, but then I get:

**POST authorization form unsuccessful after 10 attempts.**

I can't get any further - Any thoughts on a solution?

**Reply by Pyromancer579** [reply]
All sorted now for the tokens - thanks to all those who contributed.

But like @mathursharad74 I could do with being pointed at an example API call, so I can change the modes on my Powerwall, as the old code I was using doesn't seem to work with the new token system.

**Comment by mathursharad74** [comment]
Thanks for the new Oauth2 code. 
Now can someone point me to an example to where I can use the token to make an actual API call to send commands to Tesla to perform operations on my car?

Thx

**Reply by BenjiFTL** [reply]
The docs on the api for the commands should still work fine now that you have to token. https://tesla-api.timdorr.com/vehicle/commands

Do note you have to first get your vehicle's id for the proper application. It's all explained in the docs I linked.

**Reply by Pyromancer579** [reply]
Thanks @BenjiFTL , unfortunatly for me @timdorr has concentrated on Tesla Vehicle controls, where as I need to read the Mode of my Powerwall/Gateway  (backup, self_consumption or autonomous) and then set the Mode to one of these three options.

Can anyone point me at the correct Python syntax for achieving this?

**Reply by BenjiFTL** [reply]
These endpoints may still indeed work, but I have no way of knowing. Worth looking into though. https://www.teslaapi.io/powerwalls/state-and-settings
